File: LeetCode/Grind 75/two-sum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # O(N) by hash map
        d = {}
        for i, prev in enumerate(nums):
            next = target - prev
            if next in d:
                return [d[next], i] 
            
            d[prev] = i


        # O(N^2) time complexity
        # length = len(nums)
        # for i in range(length):
        #     for j in range(length):
        #         if i == j:
        #             continue

        #         if nums[i] + nums[j] == target:
        #             return [i, j]

        # return [0, 1] # impossible, because there is always a answer

File: LeetCode/Grind 75/two-sum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        // O(n^2), backtracking
        /*
         * for i in range (0, len(n))
         * for j in range(i, len(n))
         * res := nums[i] + num[j]
         * if res == target
         * return i, j
         */
        for (int i = 0; i < nums.length; i++) {
            for (int j = i + 1; j < nums.length; j++) {
                int sum = nums[i] + nums[j];
                if (sum == target) {
                    int[] result = new int[] { i, j };
                    return result;
                }
            }
        }

        return null;
    }
}